### ISSUE TITLE
alternative shellvars_list_seq lens

### ISSUE THREAD

**Issue by huppmann** [hercules-team/augeas#381]
The Shellvars_list.lns parses an assigned value into either a single value node or a list of value/* nodes depending on whether the value consists of multiple words or not.  Adding a value at a node then requires distinguishing the case when there is a single value, as just adding a subscripted value to such a node will replace the value rather than turning it into a multi-valued node.  This is particularly problematic for puppet augeas resources, because their *changes* and *only_if* parameters don't (easily?) allow handling the single- and multi-valued cases separately

The attached is just a variation of Shellvars_list.lns called Shellvars_list_seq.lns which always generates a value/* subtree, even for a single-valued assignment, thus allowing for uniform treatment of single- and multi-valued assignments.  The caveat is, if it parses a single-valued assignment **without quote delimiters**, and a value is added to it without setting a quote, then it will produce a (likely) bogus statement when it is saved.  Hence, when adding a value/*, it is advisable to set quote as well, but obviously that has its perils, if any of the values contain the quote character that you happened to pick

[Uploading shellvars_list_seq.aug.txt…]()

**Comment by huppmann** [comment]
i don't know what happened to the attachment ... here's the diff from the standard lens, inlined:

i forgot to change the `let filter =` bit, so you may have to do that, in any version that you release

i should also mention that i have no idea how ugly this is, not being an augeas-lens hacker or knowing anything about ML, so please take with a grain of salt

<pre>
--- /usr/share/augeas/lenses/dist/shellvars_list.aug	2014-06-10 12:12:58.000000000 -0400
+++ shellvars_list_seq.aug.txt	2016-06-18 12:58:18.288682000 -0400
@@ -1,7 +1,24 @@
 (* Generic lens for shell-script config files like the ones found *)
 (* in /etc/sysconfig, where a string needs to be split into       *)
 (* single words.                                                  *)
-module Shellvars_list =
+(*                                                                *)
+(* This is a variation on Shellvars_list that returns sequences   *)
+(* in all cases, rather than only if there are multiple values at *)
+(* a node. This allows for uniform handling of single- and multi- *)
+(* word assignments when adding values at a node. *However*, note *)
+(* that adding a value at a node that has no quote character is   *)
+(* going to produce output that won't parse as a shell assignment *)
+(* statement (only[*]), so you may need to be sure to set an ap-  *)
+(* propriate quote character whenever you may possibly be turning *)
+(* an unquoted single-word assignment into a multiple-word as-    *)
+(* signment.                                                      *)
+(* [*] In the Bourne shell, it will parse as an assignment pre-   *)
+(* fixed to a command invocation, with the word at augeas         *)
+(* value[2] being parsed as the command to exec (unless that con- *)
+(* tains an unquoted equal sign, producing another assignment,    *)
+(* which may in turn precede ... )                                *)
+
+module Shellvars_list_seq =
   autoload xfm
 
   let eol = Util.eol
@@ -22,21 +39,24 @@
   let list(word:regexp) =
     let list_value = store word in
       indent .
-      [ label "value" . list_value ] .
-      [ del /[ \t\n]+/ " "  . label "value" . list_value ]* . indent
+      [ seq "value" . list_value ] .
+      [ del /[ \t\n]+/ " "  . seq "value" . list_value ]* 
+      . indent
 
 
   (* handle single quoted lists *)
   let squote_arr = [ label "quote" . store /'/ ]
-                   . (list sqword)? . del /'/ "'"
+                   . [ label "value" . counter "value" . (list sqword)? . del /'/ "'" ]
 
   (* similarly handle double qouted lists *)
   let dquote_arr = [ label "quote" . store /"/ ]
-                   . (list dqword)? . del /"/ "\""
+                   . [ label "value" . counter "value" . (list dqword)? . del /"/ "\"" ]
 
   (* handle unquoted single value *)
   let unquot_val = [ label "quote" . store "" ]
-                 . [ label "value"  . store (uqword+ | bqword)]?
+                 . [ label "value"
+                 . [ counter "value" . seq "value"  . store (uqword+ | bqword)]?
+                   ]
 
 
   (* lens for key value pairs *)
</pre>

**Comment by raphink** [comment]
Do you know about https://forge.puppet.com/herculesteam/augeasproviders_shellvar ?

**Comment by huppmann** [comment]
On Sun, Jun 19, 2016 at 02:40:58AM -0700, Rapha�l Pinson wrote:
> Do you know about https://forge.puppet.com/herculesteam/augeasproviders_shellvar ?

i do, thanks, and i use a lot of your other augeas-based providers as is

the shellvar provider doesn't seem to allow adding to or removing elements
from an array individually, however, which is what i wanted. i.e., i want
to remove "quiet" and "splash" from GRUB_CMDLINE_LINUX_DEFAULT in
/etc/default/grub and make sure "nosplash" is there, for good measure, but
i don't want to remove anything else that might already be there, so i can't
just reassign it wholesale or add to it, since i want to surgically remove
stuff from it

in my case, yes, maybe hacking on the shellvar provider would be better, but
i thought having a lens that did what i want might prove more useful
generally, even if i have to hack up augeas resources directly

**Comment by raphink** [comment]
Yes, the shell provider lets you do that: https://forge.puppet.com/herculesteam/augeasproviders_shellvar/readme#appending-to-arrays

**Comment by huppmann** [comment]
On Sun, Jun 19, 2016 at 12:35:23PM -0700, Rapha�l Pinson wrote:
> Yes, the shell provider lets you do that:
> https://forge.puppet.com/herculesteam/augeasproviders_shellvar/readme#appending-to-arrays

believe me: i would be happy to use one of your (awesome) providers in
preference to coding up horrifying only_if expressions (as i am already
doing with your hosts provider, aliases, etc.), but i don't see there,
in the documentation you referenced or in

https://github.com/hercules-team/augeasproviders_shellvar

where you can *remove* a elements from an array but otherwise leave it in
tact (and/or, optionally, add some elements)

like i said, your provider is 99% awesome, especially the bit that comments
and uncomments, keeping things in sequence, so i would be happy to use it
(not that having a Shellvars_list_seq.lns, as i maybe wrongly adopted as the
solution to my problem, is without merit) but i am just missing that one
feature.  (to put a fine point on it, i want something--taking a snippet
from your documentation--like

	shellvar { "GRUB_CMDLINE_LINUX":
	  ensure       => present,
	  target       => "/etc/default/grub",
	  value        => ["quiet", "cgroup_enable=memory"],

	 #||||||||||||||||||||||||||
	 #vvvvvvvvvvvvvvvvvvvvvvvvvv

	  antivalues   => ["splash"],
	  array_modify => true,

	 #^^^^^^^^^^^^^^^^^^^^^^^^^^
	 #||||||||||||||||||||||||||
	}

so that, if "splash" is there, it is stripped.  (ignore the specifics or
grub and Linux kernel bootparams, please.  i realize maybe appending
"nosplash" would do the same, if "splash" is already there. also,
"antivalues" is obviously a throwaway term, and "modify" doesn't necessarily
express the intent of "adding and subtracting, not replacing".  i leave it
to puppet catalog UX engineers to come up with something better)

again (because i hadn't said it previously but meant to):  your providers/
types are awesome, and i would use one of them if i could in this case,
as i have in many others.  thanks for those others.  if you can modify
your shellvars provider/type to accommodate my need to remove array ele-
ments, i will withdraw my stupid lens suggestion with pleasure

**Comment by raphink** [comment]
Ok, let me get your need straight. You need to be able to control array members individually, ensuring their presence.

Maybe what you need is some kind of `shellvar_array_element` type/provider:

```puppet
shellvar_array_element { 'GRUB_CMDLINE_LINUX/quiet':
  ensure => present,
  target   => '/etc/default/grub',
}

shellvar_array_element { 'GRUB_CMDLINE_LINUX/splash':
  ensure => absent,
}
```

Would that do?

**Comment by raphink** [comment]
Now if you still want to modify the lens, could you please open a PR with a test showing what you want to achieve with that new lens?
